Tests/GRPCHTTP2CoreTests/Client/Connection/GRPCChannelTests.swift

```diff
@@ -747,6 +747,65 @@ final class GRPCChannelTests: XCTestCase {
       group.cancelAll()
     }
   }
+
+  func testQueueRequestsThenClose() async throws {
+    // Start a server.
+    let server = TestServer(eventLoopGroup: .singletonMultiThreadedEventLoopGroup)
+    let address = try await server.bind()
+    let (resolver, continuation) = NameResolver.dynamic(updateMode: .push)
+    continuation.yield(.init(endpoints: [Endpoint(address)], serviceConfig: nil))
+
+    let channel = GRPCChannel(
+      resolver: resolver,
+      connector: .posix(),
+      config: .defaults,
+      defaultServiceConfig: ServiceConfig()
+    )
+
+    try await withThrowingDiscardingTaskGroup { group in
+      group.addTask {
+        try await server.run(.echo)
+      }
+
+      group.addTask {
+        await channel.connect()
+      }
+
+      for try await state in channel.connectivityState {
+        switch state {
+        case .ready:
+          try await channel.withStream(descriptor: .echoGet, options: .defaults) { stream in
+            try await stream.outbound.write(.metadata([:]))
+
+            var iterator = stream.inbound.makeAsyncIterator()
+            _ = try await iterator.next()  // initial metadata
+
+            // Drop the connection, this will trigger a state change to transient failure.
+            server.clients.first?.close(promise: nil)
+          }
+
+        case .transientFailure:
+          group.addTask {
+            channel.close()
+          }
```

Comment:
Does this have to be inside a task?

Reply:
This test is inherently racy: we have to close after the continuation for `withStream` has been queued, but we can't order that in code (because `withStream` waits for the continuation to be resumed). 

While the channel doesn't pick up the close event until it consumes it from the async sequence, the task adds a little bit more of a delay meaning that the continuation gets queued first. However, I think we can lean into this more by adding a small sleep before calling close. I'll add that.

Reply:
I ended up reworking the test to make it a bit more reliable.